tests/ts/lsfd/mkfds-tcp

```diff
@@ -22,6 +22,22 @@ ts_init "$*"
 
 ts_check_test_command "$TS_CMD_LSFD"
 ts_check_test_command "$TS_HELPER_MKFDS"
+
+# This test cases doesn't work well on Qemu user mode emulation.
+#
+# tests_mkfds command uses hton* functions when passing
+# IP addresses and TCP ports to Linux via bind(2) and connect(2).
+# On the other hand, Linux provides these information in host
+# byte order via /proc/net/tcp that lsfd reads.
+#
+# If Qemu emulates a big-endian architecture (e.g. s390) but its
+# host Linux runs on little-endian architecture (e.g. x86_64),
+# this test case doesn't work.
+#
+# test_mkfds built for s390 passes htonl(127.0.0.1) value to
+# x86_64 Linux. The linux may receive 1.0.0.127. The Linux
+# prints "1.0.0.127" in /proc/net/tcp. lsfd reads the string.
```

Comment:
`/proc/net/tcp` using binary data in host-endianess is a bit unfortunate. Especially as there seems to be no way of getting the endianess of the kernel from `/proc`. This is also tripping up lscpu. Maybe a kernel patch would be in order.

Reply:
About /proc/net/tcp, our use case is special. So I doubt that kernel developers accept such a patch.

Reply:
I think we can try it.
@masatake or @karelzak could you provide the contents of `/proc/net/tcp` from a native s390x machine?

Reply:
OTOH, why are we treating IP addresses as numbers anyways? Shouldn't they be plain arrays of bytes?

Reply:
I could not find this idea! I have just rewritten the code based on your idea.
The new code works fine on x86_64. I will make a pull request for the change.

Reply:
It seems I misunderstood your idea. 
What I tried was:
```
diff --git a/misc-utils/lsfd-sock-xinfo.c b/misc-utils/lsfd-sock-xinfo.c
index 1ef2f40b7..20c047942 100644
--- a/misc-utils/lsfd-sock-xinfo.c
+++ b/misc-utils/lsfd-sock-xinfo.c
@@ -615,34 +615,52 @@ static void load_xinfo_from_proc_inet_L3(ino_t netns_inode, const char *proc_fil
                goto out;
 
        while (fgets(line, sizeof(line), tcp_fp)) {
-               unsigned long local_addr;
-               unsigned long local_port;
-               unsigned long remote_addr;
-               unsigned long remote_port;
+               union {
+                       uint32_t n;
+                       uint8_t b[4];
+               } local_addr, remote_addr;
+               unsigned int local_atmp[4], remote_atmp[4];
+               union {
+                       uint16_t n;
+                       uint8_t b[2];
+               } local_port, remote_port;
+               unsigned int local_ptmp[2], remote_ptmp[2];
                unsigned long st;
                unsigned long long inode;
                struct tcp_xinfo *tcp;
                struct inet_xinfo *inet;
                struct sock_xinfo *sock;
 
-               if (sscanf(line, "%*d: %lx:%lx %lx:%lx %lx %*x:%*x %*x:%*x %*x %*u %*u %lld",
-                          &local_addr, &local_port, &remote_addr, &remote_port,
-                          &st, &inode) != 6)
+               if (sscanf(line, "%*d: %2x%2x%2x%2x:%2x%2x %2x%2x%2x%2x:%2x%2x %lx %*x:%*x %*x:%*x %*x %*u %*u %lld",
+                          &(local_atmp[0]), &(local_atmp[1]), &(local_atmp[2]), &(local_atmp[3]),
+                          &(local_ptmp[0]), &(local_ptmp[1]),
+                          &(remote_atmp[0]), &(remote_atmp[1]), &(remote_atmp[2]), &(remote_atmp[3]),
+                          &(remote_ptmp[0]), &(remote_ptmp[1]),
+                          &st, &inode) != 14)
                        continue;
 
                if (inode == 0)
                        continue;
 
+               for (size_t i = 0; i < 4; i++) {
+                       local_addr.b[i] = (uint8_t)local_atmp[i];
+                       remote_addr.b[i] = (uint8_t)remote_atmp[i];
+               }
+               for (size_t i = 0; i < 2; i++) {
+                       local_port.b[i] = (uint8_t)local_ptmp[i];
+                       remote_port.b[i] = (uint8_t)remote_ptmp[i];
+               }
+
                tcp = xmalloc(sizeof(struct tcp_xinfo));
                inet = (struct inet_xinfo *)tcp;
                sock = (struct sock_xinfo *)inet;
                sock->class = class;
                sock->inode = (ino_t)inode;
                sock->netns_inode = netns_inode;
-               inet->local_addr = local_addr;
-               tcp->local_port = local_port;
-               inet->remote_addr = remote_addr;
-               tcp->remote_port = remote_port;
+               inet->local_addr = ntohl(local_addr.n);
+               tcp->local_port = ntohs(local_port.n);
+               inet->remote_addr = ntohl(remote_addr.n);
+               tcp->remote_port = ntohs(remote_port.n);
                tcp->st = st;
 
                add_sock_info(sock);
```

I run lfsd with this change on x86_64. It works fine.
But this change may have no impact on lsfd running on Qmeu big-endian user mode running on x86_64 Linux.

Reply:
We can make lsfd detect the endian gap: making a socket, binding it to 127.0.01, then inspecting the socket in /proc/net/tcp. If the socket has 1.0.0.127 as a local address, lsfd runs on an environment having the gap. However, I think this is overkill.

Reply:
The addresses in `/proc/net/tcp` are indeed in different formats on different host-endianesses. The ports are always in the same format.
This also breaks other tools like `ss` and `lsof`

Reply:
Though the change in this thread doesn't solve the byteorder issue, but I found I will be useful for implementing TCP6.